### PR TITLE
O3-1417: Use SWR for data fetching

### PR DESCRIPTION
**Pull request by anjula-sack**
## Purpose

- The purpose of this PR is to fix #[O3-1417](https://issues.openmrs.org/browse/O3-1417)

## Goals

- To improve the data fetching performance

## Approach

- Installed SWR
- Created hooks for the APIs

### CODE REVIEW

**Comment by vasharma05** [comment]
HI @anjula-sack, thanks for switching to SWR.
I want to put up how we should use SWR.

``
const fetchingBySWR(url, params){
  const {data, error, inValidating} = useSWR(url, openmrsFetch);

  const results = useMemo(() => ({
    isLoading: !data && !error,
    locations: data?.data,
    error: error, 
    
  }), [data, error, inValidating])

  return results
}
``

**Comment by vasharma05** [comment]
You can review https://github.com/openmrs/openmrs-esm-patient-management/blob/2677ceceb04d52bf3e499c799376ffabd8dd5f93/packages/esm-patient-registration-app/src/patient-registration/patient-registration-hooks.ts#L163

**Review by vasharma05 — APPROVED** [review]

File: src/components/search-by-encounters/search-by-encounters.resources.ts [hunk 1]
```diff
@@ -1,53 +1,60 @@
...
-    data: { results },
-  } = formsResp;
-  const forms: DropdownValue[] = [];
-  results.map((form: Response, index: number) => {
-    forms.push({
-      id: index,
-      label: form.display,
-      value: form.uuid,
+export const useForms = () => {
+  const { data, error } = useSWR<{
```
Comment (brandones) [inline src/components/search-by-encounters/search-by-encounters.resources.ts:12]: This too

File: src/components/search-by-encounters/search-by-encounters.resources.ts [hunk 2]
```diff
@@ -1,53 +1,60 @@
...
-    data: { results },
-  } = encounterTypesResp;
-  const encounterTypes: DropdownValue[] = [];
-  results.map((encounterType: Response, index: number) => {
-    encounterTypes.push({
-      id: index,
-      label: encounterType.display,
-      value: encounterType.uuid,
+export const useEncounterTypes = () => {
+  const { data, error } = useSWR<{
```
Comment (brandones) [inline src/components/search-by-encounters/search-by-encounters.resources.ts:39]: Also here

File: src/components/search-by-person-attributes/search-by-person-attributes.resource.ts [hunk 3]
```diff
@@ -1,28 +1,33 @@
...
  * @returns PersonAttributes
  */
-export async function getPersonAttributes(): Promise<DropdownValue[]> {
-  const personAttributesResp: FetchResponse<{
-    results: Response[];
-  }> = await openmrsFetch("/ws/rest/v1/personattributetype", {
-    method: "GET",
-  });
+export function usePersonAttributes() {
+  const { data, error } = useSWR<{
```
Comment (brandones) [inline src/components/search-by-person-attributes/search-by-person-attributes.resource.ts:12]: This too

**Review by brandones — CHANGES_REQUESTED** [review]
Just need to use [useSWRImmutable](https://swr.vercel.app/docs/revalidation#disable-automatic-revalidations) instead of `useSWR`. Easy fix; the API is the same.

File: src/cohort-builder.resource.ts [hunk 4]
```diff
@@ -25,24 +28,26 @@ export const search = async (searchParams: SearchParams) => {
...
-    data: { results },
-  } = locationsResp;
-  const locations: DropdownValue[] = [];
-  results.map((location: Response, index: number) => {
-    locations.push({
-      id: index,
-      label: location.display,
-      value: location.uuid,
+export const useLocations = () => {
+  const { data, error } = useSWR<{
```
[the remark this answers is not in the document]
Comment (anjula-sack) [reply src/cohort-builder.resource.ts:32]: so no need to use `useMemo`?

File: src/cohort-builder.resource.ts [hunk 5]
```diff
@@ -25,24 +26,23 @@ export const search = async (searchParams: SearchParams) => {
...
       label: location.display,
       value: location.uuid,
     });
   });
 
-  return locations;
+  return {
+    isLoading: !data && !error,
+    locations,
+    locationsError: error,
```
Comment (brandones) [inline src/cohort-builder.resource.ts:46]: Normally you'd just call this `error`, and let the users of the hook resolve any naming conflicts. I leave it at your discretion.

**Comment by brandones** [comment]
I seem not to have permissions to merge.

**Comment by dkayiwa** [comment]
> I seem not to have permissions to merge.

I have just fixed the repository permissions.

**Comment by brandones** [comment]
Thanks @dkayiwa !

**Pull request merged**